File: artifacts/api-server/src/routes/quizzes.ts
import { Router } from "express";
import { and, eq, inArray, sql } from "drizzle-orm";
import { z } from "zod/v4";
import {
  db,
  quizzesTable,
  quizQuestionsTable,
  languagesTable,
} from "@workspace/db";
import { asyncHandler } from "../lib/async.ts";
import { HttpError } from "../lib/errors.ts";
import { parseIntParam, parseOptionalInt } from "../lib/parse.ts";

export const quizzesRouter = Router();

const PASS_THRESHOLD = 0.7;

const quizSelect = {
  id: quizzesTable.id,
  title: quizzesTable.title,
  description: quizzesTable.description,
  articleId: quizzesTable.articleId,
  topicId: quizzesTable.topicId,
  languageId: quizzesTable.languageId,
  createdAt: quizzesTable.createdAt,
  languageName: languagesTable.name,
};

quizzesRouter.get(
  "/quizzes",
  asyncHandler(async (req, res) => {
    const languageId = parseOptionalInt(req.query.languageId);
    const topicId = parseOptionalInt(req.query.topicId);
    const articleId = parseOptionalInt(req.query.articleId);

    const where = [];
    if (languageId !== undefined) where.push(eq(quizzesTable.languageId, languageId));
    if (topicId !== undefined) where.push(eq(quizzesTable.topicId, topicId));
    if (articleId !== undefined) where.push(eq(quizzesTable.articleId, articleId));

    const rows = await db
      .select(quizSelect)
      .from(quizzesTable)
      .leftJoin(languagesTable, eq(languagesTable.id, quizzesTable.languageId))
      .where(where.length ? and(...where) : undefined)
      .orderBy(quizzesTable.id);

    if (rows.length === 0) {
      res.json([]);
      return;
    }

    const counts = await db
      .select({
        quizId: quizQuestionsTable.quizId,
        c: sql<number>`count(*)::int`,
      })
      .from(quizQuestionsTable)
      .where(inArray(quizQuestionsTable.quizId, rows.map((r) => r.id)))
      .groupBy(quizQuestionsTable.quizId);

    const countMap = new Map(counts.map((c) => [c.quizId, c.c]));

    res.json(
      rows.map((r) => ({
        ...r,
        createdAt: r.createdAt ? r.createdAt.toISOString() : null,
        questionCount: countMap.get(r.id) ?? 0,
      })),
    );
  }),
);

quizzesRouter.get(
  "/quizzes/:id",
  asyncHandler(async (req, res) => {
    const id = parseIntParam(req.params.id, "id");
    const [quiz] = await db
      .select(quizSelect)
      .from(quizzesTable)
      .leftJoin(languagesTable, eq(languagesTable.id, quizzesTable.languageId))
      .where(eq(quizzesTable.id, id));
    if (!quiz) throw new HttpError(404, "Quiz not found");
    const questions = await db
      .select()
      .from(quizQuestionsTable)
      .where(eq(quizQuestionsTable.quizId, id))
      .orderBy(quizQuestionsTable.orderIndex, quizQuestionsTable.id);
    res.json({
      ...quiz,
      createdAt: quiz.createdAt ? quiz.createdAt.toISOString() : null,
      questionCount: questions.length,
      questions,
    });
  }),
);

const attemptSchema = z.object({
  answers: z
    .array(
      z.object({
        questionId: z.number().int().positive(),
        selectedAnswer: z.number().int().min(0),
      }),
    )
    .min(1),
});

quizzesRouter.post(
  "/quizzes/:id/attempt",
  asyncHandler(async (req, res) => {
    const id = parseIntParam(req.params.id, "id");
    const { answers } = attemptSchema.parse(req.body);

    const questions = await db
      .select()
      .from(quizQuestionsTable)
      .where(eq(quizQuestionsTable.quizId, id));
    if (questions.length === 0) throw new HttpError(404, "Quiz has no questions");

    const byId = new Map(questions.map((q) => [q.id, q]));
    const results = answers.map((a) => {
      const q = byId.get(a.questionId);
      if (!q) {
        return {
          questionId: a.questionId,
          correct: false,
          correctAnswer: -1,
          explanation: "Unknown question id",
        };
      }
      return {
        questionId: a.questionId,
        correct: a.selectedAnswer === q.correctAnswer,
        correctAnswer: q.correctAnswer,
        explanation: q.explanation ?? null,
      };
    });

    const correctCount = results.filter((r) => r.correct).length;
    const totalQuestions = questions.length;
    const score = correctCount / totalQuestions;

    res.json({
      score,
      totalQuestions,
      correctCount,
      passed: score >= PASS_THRESHOLD,
      results,
    });
  }),
);

File: apps/api/app/services/quizzes.py
from __future__ import annotations

from sqlalchemy import func, select
from sqlalchemy.orm import Session

from app.models import Language, Quiz, QuizQuestion
from app.schemas import QuizAttemptOut, QuizOut, QuizQuestionOut, QuizResultItem, iso

PASS_THRESHOLD = 0.7


def _quiz_row_out(row: dict, question_count: int) -> QuizOut:
    return QuizOut(
        id=row["id"],
        title=row["title"],
        description=row.get("description"),
        article_id=row.get("article_id"),
        topic_id=row.get("topic_id"),
        language_id=row.get("language_id"),
        language_name=row.get("language_name"),
        question_count=question_count,
        created_at=iso(row.get("created_at")),
    )


def list_quizzes(
    db: Session,
    *,
    language_id: int | None = None,
    topic_id: int | None = None,
    article_id: int | None = None,
) -> list[QuizOut]:
    stmt = (
        select(
            Quiz.id,
            Quiz.title,
            Quiz.description,
            Quiz.article_id,
            Quiz.topic_id,
            Quiz.language_id,
            Quiz.created_at,
            Language.name.label("language_name"),
        )
        .outerjoin(Language, Language.id == Quiz.language_id)
        .order_by(Quiz.id)
    )
    if language_id is not None:
        stmt = stmt.where(Quiz.language_id == language_id)
    if topic_id is not None:
        stmt = stmt.where(Quiz.topic_id == topic_id)
    if article_id is not None:
        stmt = stmt.where(Quiz.article_id == article_id)

    rows = db.execute(stmt).mappings().all()
    if not rows:
        return []

    quiz_ids = [r["id"] for r in rows]
    count_map: dict[int, int] = {}
    for row in db.execute(
        select(QuizQuestion.quiz_id, func.count().label("c"))
        .where(QuizQuestion.quiz_id.in_(quiz_ids))
        .group_by(QuizQuestion.quiz_id)
    ):
        count_map[row[0]] = row[1]

    return [_quiz_row_out(dict(r), count_map.get(r["id"], 0)) for r in rows]


def get_quiz(db: Session, quiz_id: int) -> QuizOut:
    stmt = (
        select(
            Quiz.id,
            Quiz.title,
            Quiz.description,
            Quiz.article_id,
            Quiz.topic_id,
            Quiz.language_id,
            Quiz.created_at,
            Language.name.label("language_name"),
        )
        .outerjoin(Language, Language.id == Quiz.language_id)
        .where(Quiz.id == quiz_id)
    )
    row = db.execute(stmt).mappings().first()
    if not row:
        raise LookupError("Quiz not found")

    questions = db.scalars(
        select(QuizQuestion)
        .where(QuizQuestion.quiz_id == quiz_id)
        .order_by(QuizQuestion.order_index, QuizQuestion.id)
    ).all()

    out = _quiz_row_out(dict(row), len(questions))
    out.questions = [
        QuizQuestionOut(
            id=q.id,
            quiz_id=q.quiz_id,
            question=q.question,
            options=q.options,
            correct_answer=q.correct_answer,
            explanation=q.explanation,
            order_index=q.order_index,
        )
        for q in questions
    ]
    return out


def submit_quiz_attempt(
    db: Session,
    quiz_id: int,
    answers: list[dict],
) -> QuizAttemptOut:
    questions = db.scalars(select(QuizQuestion).where(QuizQuestion.quiz_id == quiz_id)).all()
    if not questions:
        raise LookupError("Quiz has no questions")

    by_id = {q.id: q for q in questions}
    results: list[QuizResultItem] = []
    for a in answers:
        q = by_id.get(a["question_id"])
        if not q:
            results.append(
                QuizResultItem(
                    question_id=a["question_id"],
                    correct=False,
                    correct_answer=-1,
                    explanation="Unknown question id",
                )
            )
        else:
            results.append(
                QuizResultItem(
                    question_id=a["question_id"],
                    correct=a["selected_answer"] == q.correct_answer,
                    correct_answer=q.correct_answer,
                    explanation=q.explanation,
                )
            )

    correct_count = sum(1 for r in results if r.correct)
    total = len(questions)
    score = correct_count / total
    return QuizAttemptOut(
        score=score,
        total_questions=total,
        correct_count=correct_count,
        passed=score >= PASS_THRESHOLD,
        results=results,
    )
